feat: enable easy-opening of populated SSO authentication URLs

brassard opened the pull request:
Parse authentication messages and automatically open SSO URLs with user codes, eliminating manual copy/paste workflow.

- Add `handle_sso_msg()` function to augment `window/showMessage` for SSO login
  - Extracts URL and `user_code` from SSO message
  - Prepares populated URL with `?user_code=xxxx-xxxx` parameter
  - Injects `<CR>` keybinding interaction into popup to open populated URL
- Add `util.open()` function for standardized URL opening
- Add `util.with_modifiable()` helper for buffer modifications

### Description of changes:

These changes to enhance the login experience. The prompt to visit a URL and copy/paste a code is relatively rare, but when required, it is a bit annoying to jockey around the prompt string with keyboard to collect the code and open the URL link. (i.e. `$bbbv$y/http<CR>gx`)

Initially I thought about trying to reformat the string to make it more easily navigable at least … but then I discovered that I was able to pass the code as a URL search parameter (`user_code`) like so…
```
https://view.awsapps.com/start/#/device?user_code=MY_CODE_XYZ
```

So, we can intercept matching messages and then provide a way to quickly open that URL. 

My first implementation auto-opened the URL, but that was some mix of surprising/overwhelming/annoying. After some iteration, I landed on enabling a keymap equivalent to `<CR> => vim.ui.open()` in the popup window.

### Upgraded flow:
1. Prompt displayed
1. User types ENTER (`<CR>`) _**<< NEW INTERACTION**_
1. URL is opened in browser
    1. If no browser is present, system messages that URL failed to open
1. Browser loads populated `awsapps` URL
1. User confirms codes match and accepts permissions
1. User returns to console as fulfilled prompt exits as normal


https://github.com/user-attachments/assets/a7141b8b-c1b6-4d99-907f-fc48908f4eec

> By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

justinmk3 committed:
docs: fix anchor name (#5)

brassard-aws committed:
feat: enable easy-opening of populated SSO authentication URLs

Parse authentication messages and automatically open SSO URLs with
user codes, eliminating manual copy/paste workflow.

- Extract URL and user code from LSP showMessage
- Prepare combined URL with ?user_code= parameter
- Add <Enter> keymap to popup for automated open
- Add util.open() function for standardized URL opening
- Add util.with_modifiable() helper for buffer modifications

justinmk3 commented on lua/amazonq/lsp.lua:
```diff
@@ -61,6 +62,39 @@ M.config --[[@type vim.lsp.ClientConfig]] = {
+--- @param result table LSP showMessage result
+--- @param popup_win integer Popup window handle
+--- @return boolean true if SSO behavior was injected
+function M.handle_sso_msg(result, popup_win)
```
`on_` convention is used everywhere in this codebase. `handle_` is not used in this codebase.

brassard replied:
Fixed.

justinmk3 commented on lua/amazonq/lsp.lua:
```diff
@@ -61,6 +62,39 @@ M.config --[[@type vim.lsp.ClientConfig]] = {
+--- @param result table LSP showMessage result
+--- @param popup_win integer Popup window handle
+--- @return boolean true if SSO behavior was injected
+function M.handle_sso_msg(result, popup_win)
```
```suggestion
function M.on_sso_msg(result, popup_win)
```

brassard replied:
Fixed.

justinmk3 commented on lua/amazonq/util.lua:
```diff
@@ -217,4 +217,25 @@ function M.indent(indent_width, lines)
+  local was_modifiable = vim.bo[buf].modifiable
+  vim.bo[buf].modifiable = true
+  fn()
+  vim.bo[buf].modifiable = was_modifiable
```
this is a rather special-case function, only used once. would be clearer to just set modifiable directly instead. or else generalize this.

also, if `fn()` raises an error then 'modifiable' is not restored. use `pcall(fn, ...)`

brassard replied:
Removed function, moved logic and also wrapped the main function in `pcall(...)` in the usage location.

justinmk3 commented on lua/amazonq/util.lua:
```diff
@@ -217,4 +217,25 @@ function M.indent(indent_width, lines)
+  if success then
+    M.msg('Opened: ' .. path)
+  else
+    M.msg('Failed to open: ' .. path)
```
why not use `vim.ui.open` directly? it already shows an error when needed.

brassard replied:
Mostly just wasn't familiar with whether it could be depended on to do the right thing.

Removed extra function and now calling `vim.ui.open` directly.

Thanks for the pointer.

justinmk3 commented on lua/amazonq/lsp.lua:
```diff
@@ -61,6 +62,39 @@ M.config --[[@type vim.lsp.ClientConfig]] = {
+  -- For popup buffer only, add keymap to open url
+  vim.keymap.set('n', '<CR>', function()
+    util.open(auth_url)
+  end, { buffer = buf, nowait = true, silent = true })
```
Just for background: we previously had a modal dialog, and it would open the URL without needing mappings, etc.

The reason for using a floating window is because `vim.ui.open` does not work on remote (ssh) systems. In those systems, the user needs to copy the URL and put it in their browser.

brassard replied:
What do you think about adding a keymap to yank the URL too?

`<CR>` => open
`y` => copy

justinmk3 replied:
No, users already know how to copy text. Adding random `y` mappings just adds a layer of uncertainty that must be learned, and often will not be wanted.

A `<CR>` mapping is low-risk and intuitive, so no objection to that.

justinmk3 commented on lua/amazonq/lsp.lua:
```diff
@@ -61,6 +62,39 @@ M.config --[[@type vim.lsp.ClientConfig]] = {
+  end, { buffer = buf, nowait = true, silent = true })
+  -- Update the popup content to include instructions
+  util.with_modifiable(buf, function()
+    vim.api.nvim_buf_set_lines(buf, 1, 1, false, { '', 'Press <Enter> to open URL in browser' })
```
Instead of modifying the buffer, why not modify the message (or create a new message), before `M.fmt_msg` is called? That avoids needing to manage 'modifiable', and lets presentation decisions live in one place.

brassard replied:
I was trying to add the keymaps and text changes all in one shot as injections, and only when appropriate.

Another path might be to change `show_popup()` logic to pass in keymaps, but I didn't feel comfortable changing much existing logic.

brassard replied:
I'm going to iterate on this a little bit... I think it can be restructured a little bit to be closer to what you are describing.

brassard replied:
@justinmk3 I cooked a variation that adds an optional `default_action` function param to `show_popup()` that will get mapped to `<CR>` if present. 

If a populated `auth_url` can be extracted, then the `window/showMessage` handler will modify the message to indicate that the user can type \<Enter\> to open the URL in their browser and provide the appropriate `open_auth_url` function.

Code below ... I will push this commit for review/iteration if you think this is an improved approach over the hackier buffer modifications.

```diff
diff --git a/lua/amazonq/lsp.lua b/lua/amazonq/lsp.lua
index 677efea..a1de382 100644
--- a/lua/amazonq/lsp.lua
+++ b/lua/amazonq/lsp.lua
@@ -42,7 +42,26 @@ M.config --[[@type vim.lsp.ClientConfig]] = {
     -- Handle window events - this is used to display the authentication url and code to the user
     ['window/showMessage'] = function(_, result, _ctx)
       if result and result.message then
-        util.show_popup(M.fmt_msg(result))
+        local open_auth_url = nil
+
+        -- Extract auth_url if it exists.
+        local auth_url = util.extract_auth_url(result.message)
+
+        -- If the auth_url is found, prepare default actions for the popup.
+        if auth_url then
+          -- Define the URL opening function.
+          open_auth_url = function()
+            vim.ui.open(auth_url)
+          end
+          -- Add user note about URL opening.
+          result = vim.tbl_extend(
+            'force',
+            result,
+            { message = result.message .. '\n\n' .. 'Press <Enter> to open URL in browser' }
+          )
+        end
+
+        util.show_popup(M.fmt_msg(result), nil, nil, open_auth_url)
       end
     end,
   },
diff --git a/lua/amazonq/util.lua b/lua/amazonq/util.lua
index c8f76dc..c0ef121 100644
--- a/lua/amazonq/util.lua
+++ b/lua/amazonq/util.lua
@@ -54,8 +54,9 @@ end
 --- @param text string
 --- @param max_width? integer
 --- @param max_height? integer
+--- @param default_action? function The function executed on <CR>.
 --- @return number # window-id
-function M.show_popup(text, max_width, max_height)
+function M.show_popup(text, max_width, max_height, default_action)
   M.dismiss_popup() -- Only show one "dialog" at any given time.
   assert(text)

@@ -74,6 +75,15 @@ function M.show_popup(text, max_width, max_height)
     relative = 'editor',
     title = M.productName,
   })
+
+  -- For popup buffer only, add <CR> keymap to execute default_action if it exists.
+  local buf = vim.api.nvim_win_get_buf(win)
+  if default_action then
+    vim.keymap.set('n', '<CR>', function()
+      pcall(default_action)
+    end, { buffer = buf, nowait = true, silent = true })
+  end
+
   vim.api.nvim_set_current_win(win)
   popup_win = win
   return win
@@ -217,4 +227,19 @@ function M.indent(indent_width, lines)
   return type(lines) == 'string' and table.concat(indented, '\n') or indented
 end

+--- Extracts SSO authentication URL from a server response string.
+--- @param message string LSP response message.
+--- @return string|nil The URL to be opened.
+function M.extract_auth_url(message)
+  -- This regex pattern is very specific to the existing display, but will be ignored if no match is found.
+  local base_url, user_code = message:match(
+    '^To proceed, open the login page (https://[^%s]+) and provide this code to confirm the access request: ([A-Z0-9%-]+)$'
+  )
+  if base_url and user_code then
+    return base_url .. '?user_code=' .. user_code
+  else
+    return nil
+  end
+end
+
 return M
```

justinmk3 replied:
> a variation that adds an optional `default_action` function param to `show_popup()` that will get mapped to `<CR>` if present.

That increases the net code cost and abstracts "creating a mapping" which is not really something that needs abstraction, because keymap creation is a common and lost-cost task in vim. Also, that approach spreads out the logic instead of keeping it isolated.

The purpose of this PR is a workaround to add "modal" behavior to a non-modal dialog. The reason the dialog is non-modal is because users need to copy-paste from it when running in a remote (ssh) env. This is a special case, not something that needs ergonomic encouragement. If a modal dialog were needed then show_popup() wouldn't be needed. Instead the normal vim mechanisms such as `vim.fn.input()` or `vim.ui.input()` could be used.

brassard-aws committed:
feat_openSsoUrl: responding to PR feedback

viktorsaws commented:
Could you change target branch to `develop` instead of `main`? Currently we're using `develop` for staging all changes.

justinmk3 commented on lua/amazonq/lsp.lua:
```diff
@@ -61,6 +62,39 @@ M.config --[[@type vim.lsp.ClientConfig]] = {
+  end, { buffer = buf, nowait = true, silent = true })
+  -- Update the popup content to include instructions
+  vim.bo[buf].modifiable = true
+  pcall(vim.api.nvim_buf_set_lines, buf, 1, 1, false, { '', 'Press <Enter> to open URL in browser' })
```
why use pcall()  ? this should never fail. if it does, showing an error will lead to a bugfix.

brassard replied:
This is here due to previous feedback and then I missed pulling across the `was_modifiable` state of the buffer.

I added this back in for the latest, which makes the `pcall()` version useful to ensure restored modifiable state.

justinmk3 commented on lua/amazonq/lsp.lua:
```diff
@@ -61,6 +62,39 @@ M.config --[[@type vim.lsp.ClientConfig]] = {
+--- @param result table LSP showMessage result
+--- @param popup_win integer Popup window handle
+--- @return boolean true if SSO behavior was injected
+function M.on_sso_msg(result, popup_win)
```
This probably fits in the util.lua module, just below this: https://github.com/awslabs/amazonq.nvim/blob/814e2c66d82cb1b5f5107fbd3bccfae63d120142/lua/amazonq/util.lua#L83

That's already where popup-related logic lives. And this doesn't have much relevance to `lsp.lua`, it's more like a temporary hack.

brassard replied:
Moved ALL changes to util.lua, where it's now only updating the text that gets passed into the popup.

justinmk3 reviewed (approved):
LGTM after comments are addressed. I'm not sure this is a great practice to follow or repeat in the future, but it solves a near-term problem in an isolated way that can be removed/revisited later with low risk.

brassard commented on lua/amazonq/lsp.lua:
```diff
@@ -61,6 +62,39 @@ M.config --[[@type vim.lsp.ClientConfig]] = {
+  local buf = vim.api.nvim_win_get_buf(popup_win)
+  -- For popup buffer only, add keymap to open url
+  vim.keymap.set('n', '<CR>', function()
+    pcall(vim.ui.open, auth_url)
```
@justinmk3 I can remove this `pcall()` as well, right?

brassard replied:
Ironically, it seems this one *can* be removed and the other one is supposed to be there, once I added the `was_modifiable` handling back in.

brassard-aws committed:
feat_openSsoUrl: responding to PR feedback

* isolating logic to `util.lua`
* modifying pcall() behavior

brassard commented:
@justinmk3 I moved ALL changes completely into `util.lua`.

1. Now, the text in the shared `popup_win` is checked and modified when the regex matches. It continues to be hacky, but is now completely specific to the `show_popup()` logic. Given you wanted it to live with `show_popup()` / `dismiss_popup()`, it felt natural to just do all the processing there for now.
2. I exposed the populated link directly in the popup for simplified handling, even for those that can't open a browser (see screenshot). <img width="820" height="141" alt="Screenshot 2025-09-10 at 07 57 57" src="https://github.com/user-attachments/assets/043b0d4f-ac97-4cfd-8c8d-ee721a5ab282" />

Let me know how you feel about 1 & 2. Thanks.

justinmk3 commented on lua/amazonq/util.lua:
```diff
@@ -89,6 +93,39 @@ function M.dismiss_popup()
+--- Injects SSO authentication behavior into an existing popup window
+--- @param message string The popup message.
+--- @return boolean true if SSO behavior was injected
+function M.on_sso_msg(message)
```
Not a blocker: I know I suggested `on_` before, but now that I understand this better it seems clear that it's just attempting to format stuff, so `try_format_...` is a good name.


```suggestion
function M.try_format_sso_msg(message)
```

brassard replied:
Yea ... I was resisting changing it to avoid churn, but I like your name suggestion.

Changed in the latest version.

brassard-aws committed:
feat_openSsoUrl: responding to PR feedback

* adjusting name of SSO message formatter function
